src/backend/distributed/commands/extension.c

```diff
@@ -491,6 +495,114 @@ PreprocessAlterExtensionUpdateStmt(Node *node, const char *queryString)
+	 */
+	if (!IsCoordinator())
+	{
+		return;
+	}
+
+	bool citusIsUpdatedToLatestVersion = InstalledAndAvailableVersionsSame();
+
+	/*
+	 * Knowing that Citus version was different than the available version before
+	 * standard process utility runs ALTER EXTENSION command, we perform post
+	 * process operations if Citus is updated to that available version
+	 */
+	if (!citusIsUpdatedToLatestVersion)
+	{
+		return;
+	}
+
+	/*
+	 * Finally, mark existing objects that are not included in distributed object
+	 * infrastructure in older versions of Citus (but now should be) as distributed
+	 */
+	MarkExistingObjectDependenciesDistributedIfSupported();
+}
+
+
+/*
+ * MarkAllExistingObjectsDistributed marks all objects that could be distributed by
+ * resolving dependencies of "existing distributed tables" and "already distributed
+ * objects" to introduce the objects created in older versions of Citus to distributed
+ * object infrastructure as well.
+ *
+ * Note that this function is not responsible for ensuring if dependencies exist on
+ * nodes and satisfying these dependendencies if not exists, which is already done by
+ * EnsureDependenciesExistOnAllNodes on demand. Hence, this function is just designed
+ * to be used when "ALTER EXTENSION citus UPDATE" is executed.
+ * This is because we want to add existing objects that would have already been in
+ * pg_dist_object if we had created them in new version of Citus to pg_dist_object.
+ */
+static void
+MarkExistingObjectDependenciesDistributedIfSupported()
+{
+	ListCell *listCell = NULL;
+
+	/* resulting object addresses to be marked as distributed */
+	List *resultingObjectAddresses = NIL;
+
+	/* resolve dependencies of distributed tables */
+	List *distributedTableOidList = DistTableOidList();
+
+	foreach(listCell, distributedTableOidList)
+	{
+		Oid distributedTableOid = lfirst_oid(listCell);
+
+		ObjectAddress tableAddress = { 0 };
+		ObjectAddressSet(tableAddress, RelationRelationId, distributedTableOid);
+
+		List *distributableDependencyObjectAddresses =
+			GetDistributableDependenciesForObject(&tableAddress);
+
+		resultingObjectAddresses = list_concat(resultingObjectAddresses,
+											   distributableDependencyObjectAddresses);
+	}
+
+	/* resolve dependencies of the objects in pg_dist_object*/
+	List *distributedObjectAddressList = GetDistributedObjectAddressList();
+
+	foreach(listCell, distributedObjectAddressList)
+	{
+		ObjectAddress *distributedObjectAddress = (ObjectAddress *) lfirst(listCell);
+
+		List *distributableDependencyObjectAddresses =
+			GetDistributableDependenciesForObject(distributedObjectAddress);
+
+		resultingObjectAddresses = list_concat(resultingObjectAddresses,
+											   distributableDependencyObjectAddresses);
+	}
+
+	/* reorder object addresses list according to dependency order */
+	List *sortedObjectAddresses = OrderObjectAddressListInDependencyOrder(
+		resultingObjectAddresses);
+
+	/* finally, mark the rest of the objects as distributed */
```

Comment:
This comment is slightly stale. We're idempotently marking all objects as distributed.